src/decomposition.jl

```diff
@@ -0,0 +1,139 @@
+"""
+    decompose_zx(c)
+
+Decompose `c` into terms of Z- and X-basis.
+"""
+decompose_zx(::TT) where TT = error("`decompose_zx` is not defined for type $TT")
+decompose_zx(bir::BlockIR) = BlockIR(bir.parent, bir.nqubits, decompose_zx(bir.circuit))
+function decompose_zx(c::Chain)
+    new_chain = Chain()
+    for g in c.args
+        cg = decompose_zx(g)
+        if cg isa Chain
+            for gg in cg.args
+                push!(new_chain.args, gg)
+            end
+        else
+            push!(new_chain.args, cg)
+        end
+    end
+    return new_chain
+end
+decompose_zx(a::AdjointOperation) = AdjointOperation(decompose_zx(a.parent))
+function decompose_zx(g::Gate)
+    op = g.operation
+    loc = g.locations
+    @switch op begin
+        @case ::YGate
+            return Chain(Gate(YaoHIR.S', loc), Gate(YaoHIR.X, loc), Gate(YaoHIR.S, loc))
+        @case ::Ry
+            return Chain(Gate(YaoHIR.S', loc), Gate(Rx(theta), loc), Gate(YaoHIR.S, loc))
+        @case _
+            return g
+    end
+end
+function decompose_zx(cg::Ctrl)
+    ctrl = cg.ctrl
+    g = cg.gate
+    if length(ctrl) == 1
+        if length(g.locations) == 1
+            @switch g begin
+                @case Gate(::YGate, loc)
+                    return Chain(Gate(YaoHIR.S', loc), Ctrl(Gate(YaoHIR.X, loc), ctrl), Gate(YaoHIR.S, loc))
```

Comment:
no need to write the module name inside YaoHIR module

Reply:
It will get errors if there is no YaoHIR. Maybe it is an issue of `@switch` and `@case`.